contracts/core/libraries/QFHelper.sol

```diff
@@ -0,0 +1,120 @@
+    /// @param funder The address of the funder
+    struct Donation {
+        uint256 amount;
+        address funder;
+    }
+
+    /// Struct that defines the state of the donations to recipients
+    /// @param recipients The set of recipients
+    /// @param donations The donations for each recipient
+    struct State {
+        EnumerableSet.AddressSet recipients;
+        mapping(address => Donation[]) donations;
+    }
+
+    /// @notice Calculate the square root of a number (Babylonian method)
+    /// @param _x The number
+    /// @return _y The square root
+    function _sqrt(uint256 _x) internal pure returns (uint256 _y) {
+        uint256 _z = (_x + 1) / 2;
+        _y = _x;
+        while (_z < _y) {
+            _y = _z;
+            _z = (_x / _z + _z) / 2;
+        }
+    }
+
+    /// @notice Votes for recipients by donating
+    /// @param _state The state of the donations
+    /// @param _recipients The recipients to donate to
+    /// @param _amounts The amounts to donate to each recipient
+    /// @param _funder The address of the funder
+    /// @dev The number of recipients and amounts should be equal and the same index
+    ///      should correspond to the same recipient and amount
+    function fund(State storage _state, address[] memory _recipients, uint256[] memory _amounts, address _funder)
+        internal
+    {
+        /// Check if the number of recipients and amounts are equal
+        if (_recipients.length != _amounts.length) revert QFHelper_LengthMissmatch();
+
+        for (uint256 i = 0; i < _recipients.length; i++) {
+            /// Add the recipient to the set if it doesn't exist
+            if (!_state.recipients.contains(_recipients[i])) {
+                _state.recipients.add(_recipients[i]);
+            }
+            /// Add the donation to the recipient
+            _state.donations[_recipients[i]].push(Donation({amount: _amounts[i], funder: _funder}));
+        }
+    }
+
+    /// @notice Calculates the matching amount for each recipient using the Quadratic Funding formula
+    /// @param _state The state of the donations
+    /// @param _matchingAmount The total matching amount
+    /// @return _recipients The recipients
+    /// @return _amounts The matching amount for each recipient
+    function calculateMatching(State storage _state, uint256 _matchingAmount)
+        internal
+        view
+        returns (address[] memory _recipients, uint256[] memory _amounts)
+    {
+        /// Get the number of recipients
+        uint256 _numRecipients = _state.recipients.length();
+        /// Initialize the arrays
+        _recipients = new address[](_numRecipients);
+        _amounts = new uint256[](_numRecipients);
+
+        uint256[] memory _donationsSum = new uint256[](_numRecipients);
+        uint256 _totalContributions;
+        uint256 _sumOfSquareRoots;
+        /// Calculate the matching amount for each recipient
+        for (uint256 i = 0; i < _numRecipients; i++) {
```

Comment:
If the number of recipients is too high, iterating through all of them could become very expensive and end in a DOS condition

Reply:
i would say yes, since the calculations are just to expensive. I looked into it in the past and couldn't find any good solution.
Maybe there is a third party which we can utilize, like chainlink?

Reply:
So this was a 2 rounds ago https://gov.gitcoin.co/t/gg19-program-round-results/17281/6
It is very realistic to have 1k + unique donors who would have allocated to 1 or more projects.

I do agree with kurt that we would hit our limit pretty quick.
That's why we originally had to do it off-chain.
Additonally for Gitcoin , we end up using passport off-chain service to get scores to assign weights to matching (all which cannot be done on-chain)

At gitcoin would need the off-chain calculation. Even if we keep it as a helper for other to use , they might run it issues like running into gas limit and that feels clunky 

Waddya think @0xOneTony ?